web/blog/parquet-and-feather-writing-security-telemetry/index.qmd

```diff
@@ -0,0 +1,868 @@
+```
+
+Note that Zeek's tab-separated value (TSV) format is already a structured table,
+whereas Suricata data needs to be demultiplexed first through the `event_type`
+field.
+
+The PCAP packet type is currently hard-coded in VAST's PCAP plugin:
+
+```cpp
+type make_packet_type() {
+  const auto port_type = type{"port", count_type{}};
+  const auto timestamp_type = type{"timestamp", time_type{}};
+  return type{
+    "pcap.packet",
+    record_type{
+      {"time", timestamp_type},
+      {"src", address_type{}},
+      {"dst", address_type{}},
+      {"sport", port_type},
+      {"dport", port_type},
+      {"vlan",
+       record_type{
+         {"outer", count_type{}},
+         {"inner", count_type{}},
+       }},
+      {"community_id", type{string_type{}, {{"index", "hash"}}}},
+      {"payload", type{string_type{}, {{"skip"}}}},
+    },
+  };
+}
+```
+
+Now that we've looked at the structure of the dataset, let's take a look at our
+measurement methodology.
+
+### Measurement
+
+Our objective is understanding the storage and runtime characteristics of
+Parquet and Feather on the provided input data. To this end, we instrumented
+VAST to produce us with a measurement trace file that we then analyze with R for
+gaining insights. The [corresponding patch][patch] is not meant for further
+production, so we kept it separate. But we did find an opportunity to improve
+VAST and [made the Zstd compression level configurable][pr-2623]. Our [benchmark
+sript](benchmark.fish) is available for full reproducibility.
+
+[pr-2623]: https://github.com/tenzir/vast/pull/2623
+[patch]: feather-parquet-zstd-experiments.diff
+
+Our instrumentation produced a [CSV file](data.csv) with the following features:
+
+- **Store**: the type of store plugin used in the measurement, i.e., `parquet`
+  or `feather`.
+- **Construction time**: the time it takes to convert Arrow record batches into
+  Parquet or Feather. We fenced the corresponding code blocks and computed the
+  difference in nanoseconds.
+- **Input size**: the number of bytes that the to-be-converted record batches
+  consume.
+- **Output size**: the number of bytes that the store file takes up.
+- **Number of events**: the total number of events in all input record batches
+- **Number of record batches**
+- **Schema**: the name of the schema; there exists one store file per schema
+- **Zstd compression level**: the applied Zstd compression level
+
+Every row corresponds to a single store file where we varied some of these
+parameters. We used [hyperfine](https://github.com/sharkdp/hyperfine) as
+benchmark driver tool, configured with 8 runs. Let's take a look at the data.
```

Comment:
How did you invoke Hyperfine? The tool has a lot of configuration options. Did you configure warm-up runs? Did you clean the database after every run? This is all valuable information that I think should be included for reproducibility.

Reply:
See the adjacent `benchmark.fish` file in this PR.

Reply:
Can you link to it here?

Reply:
It's linked in the previous paragraph. I think it follows naturally when reading this in a flow, which is admittedly different when pausing in review mode.

From my perspective, it's taken care of. If you agree, please resolve.

Reply:
I'd prefer the link in this paragraph directly, but ultimately I have no strong opinion on this. Feel free to resolve.